pfSense-pkg-API/files/etc/inc/api/models/APISystemCRLCreate.inc

```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    private function __validate_descr() {
+        # Validate the required 'descr' field
+        if (isset($this->initial_data['descr'])) {
+            # Ensure description does not contain invalid characters
+            if (preg_match("/[\?\>\<\&\/\\\"\']/", $this->initial_data['descr'])) {
+                $this->errors[] = APIResponse\get(1037);
+            } else {
+                $this->validated_data["descr"] = $this->initial_data['descr'];
+            }
+        } else {
+            $this->errors[] = APIResponse\get(1002);
+        }
+    }
+
+    private function __validate_crl_data() {
+        # Validate the required 'crl_data' field
+        if (isset($this->initial_data['crl_data'])) {
+            $this->validated_data['text'] = base64_encode($this->initial_data['crl_data']);
+        } else {
+            $this->errors[] = APIResponse\get(6030);
+        }
+    }
+
+    private function __validate_lifetime() {
+        # Validate the optional 'lifetime' field
+        if (isset($this->initial_data['lifetime'])) {
+            # Ensure lifetime is within range
+            if (is_numeric($this->initial_data["lifetime"]) and intval($this->initial_data["lifetime"]) < cert_get_max_lifetime()) {
+                $this->validated_data["lifetime"] = strval($this->initial_data["lifetime"]);
+            } else {
+                $this->errors[] = APIResponse\get(1046);
+            }
+        } else {
+            $this->validated_data["lifetime"] = strval(9999);
+        }
+    }
+
+    private function __validate_serial() {
+        # Check for our optional 'serial' field
+        if (isset($this->initial_data["serial"])) {
+            # Ensure the serial number is 1 or greater
+            if (is_numeric($this->initial_data["serial"]) and intval($this->initial_data["serial"]) >= 0) {
+                $this->validated_data["serial"] = strval($this->initial_data["serial"]);
+            } else {
+                $this->errors[] = APIResponse\get(1033);
+            }
+        } else {
+            $this->validated_data["serial"] = strval(0);
+        }
+    }
+
+    public function get_all_ca() {
+        # Local variables
+        $ca_enties = [];
+
+        # Loop through each CA and add all CAs that have a valid key
+        foreach ($this->config['ca'] as $ca) {
+            if ($ca['prv']) {
+                $ca_enties[] = $ca['refid'];
+            }
+        }
+
+        return $ca_enties;
+    }
+
+    public function validate_payload() {
+        # Validate shared fields
+        $this->validated_data["refid"] = uniqid();
+        $this->__validate_method();
+        $this->__validate_caref();
+        $this->__validate_descr();
+
+        # Validate 'existing' method fields
+        if ($this->cert_method === "existing") {
+            $this->__validate_crl_data();
+        }
+
+        # Validate 'internal' method fields
+        if ($this->cert_method === "internal") {
```

Comment:
If internal CRLs are going to be supported, it looks like we are missing the fields to add the included `cert` values to the CRL. The `cert` field should probably expect an array of cert IDs and then the full configuration could be determined by the backend. You can see how pfSense does this here:

https://github.com/pfsense/pfsense/blob/e696b0868a495af4f19505b8261f25d6604adc8d/src/usr/local/www/system_crlmanager.php#L114-L164

If this isn't going to be fully supported, I would remove the `internal` method validation until it can be implemented or at least make note in the documentation that internal certificates can't be added to the CRL via API yet.

Reply:
**cert** field only applicable for updating already existing CRL, when creating new CRL i don't think it should be expected to add **cert** to CRL all in one go. Same functionality as creating CRL from web gui.
Maybe a note in documentation should be added to note that internal refer to creating new CRL.
Correct me if i am wrong.

Reply:
You are correct. My concern was mostly that users will have no way to interface with the CRL via API after it's created other than deleting it. However, for the sole purpose of just creating a CRL that can then be managed via UI this is fine for now.